___

<h1> Enunciado do Trabalho Prático (parte 2) </h1>

___
<h2> Machine Learning </h2>
<h3>  Systems Engineering and Computer Technologies </h3>
<h4> Instituto Superior de Engenharia / Universidade do Algarve </h4>

[MEEC](https://ise.ualg.pt/en/curso/1477) / [ISE](https://ise.ualg.pt) / [UAlg](https://www.ualg.pt)

Pedro J. S. Cardoso (pcardoso@ualg.pt)

___

# Previsão Binária do Status de Fumante usando Bio-Sinais

Considere a competição do Kaggle "_Binary Prediction of Smoker Status using Bio-Signals: Playground Series - Season 3, Episode 24_", que pode ser encontrada em https://www.kaggle.com/competitions/playground-series-s3e24

1. O trabalho consiste em 
    1. Comece por ler a documentação da competição.

    2. De seguida, construa um modelo para prever o _target_. Como restrição, só pode usar algoritmos da seguinte lista:
        - Regressão Logística
        - K-_Nearest Neighbors_
        - Árvores de Decisão
        - _Random Forest_
        - _Suport Vector Machines_
        - Redes Neurais

    3. Tente ajustar seu modelo o melhor possível. Sugestão: use _cross validation_ para avaliar o seu modelo e _grid search_ para ajustá-lo. 
    4. Submeta os seus resultados na plataforma do Kaggle. Lembre-se de que você só pode submeter 5 vezes por dia os seus resultados na plataforma. Note ainda que deve enviar os resultados de `predict_proba`, não do `predict`.

3. Entregue:
    - em formato Jupyter Notebook (`.ipynb`), com o nome `TP1_<axxxxx>-<nome1>_<axxxxx>-<nome2>.ipynb`, em que `<axxxxx>-<nome1>` e `<axxxxx>-<nome2>` são os número de alunos e respetivos nomes.
    - com todos os outputs, gráficos visíveis, comentários, etc.
    - com o código a funcionar, i.e., deve ser possível executar todas as células do notebook sem erros, de forma sequencial.
    - Uma captura de ecrã da sua melhor pontuação no quadro de líderes.
        
4. No final haverá uma sessão de perguntas e respostas sobre o trabalho. Será atribuida classificação negativa trabalhos que:
    - sejam cópias uns dos outros (0 valores - sem possibilidade de melhoria)
    - não sejam realizados pelos alunos que os entregam (0 valores - sem possibilidade de melhoria)
    - não sejam entregues até à data limite (desconto de 0.5 valores por dia de atraso, limite 4 valores)
    - não sejam apresentados na sessão de perguntas e respostas (0 valores)


5. Os notebooks com as melhores pontuações de teste serão apresentados em sala de aula e o(s) autor(es) receberão pontos extras.

6. Este TP seguirá o mesmo prazo da competição: 13 de novembro.

# #TODO:
 Regressão Logística DONE
 K-_Nearest Neighbors_ DONE
 Árvores de Decisão DONE
 _Random Forest_ DONE
 _Suport Vector Machines_ DONE
 Redes Neurais DONE
 Falta:
 Fazer um esemble com TUDO ( e guardar o actual que deu o melhor resultado)

# Setup

In [1]:
import pandas as pd

# load the main data
df = pd.read_csv('datasets/train.csv')
# set a random state because of ze germans
random_state=42
# show the first 5 rows
df.head()

id  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0   0   55         165          60       81.0             0.5   
1   1   70         165          65       89.0             0.6   
2   2   20         170          75       81.0             0.4   
3   3   35         180          95      105.0             1.5   
4   4   30         165          60       80.5             1.5   

   eyesight(right)  hearing(left)  hearing(right)  systolic  ...  HDL  LDL  \
0              0.6              1               1       135  ...   40   75   
1              0.7              2               2       146  ...   57  126   
2              0.5              1               1       118  ...   45   93   
3              1.2              1               1       131  ...   38  102   
4              1.0              1               1       121  ...   44   93   

   hemoglobin  Urine protein  serum creatinine  AST  ALT  Gtp  dental caries  \
0        16.5              1               1.0   22   25   27              0   
1        16.2              1               1.1   27   23   37              1   
2        17.4              1               0.8   27   31   53              0   
3        15.9              1               1.0   20   27   30              1   
4        15.4              1               0.8   19   13   17              0   

   smoking  
0        1  
1        0  
2        1  
3        0  
4        1  

[5 rows x 24 columns]

In [2]:
# prepare the data
data = df.drop("smoking", axis=1)
data.head()

id  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0   0   55         165          60       81.0             0.5   
1   1   70         165          65       89.0             0.6   
2   2   20         170          75       81.0             0.4   
3   3   35         180          95      105.0             1.5   
4   4   30         165          60       80.5             1.5   

   eyesight(right)  hearing(left)  hearing(right)  systolic  ...  \
0              0.6              1               1       135  ...   
1              0.7              2               2       146  ...   
2              0.5              1               1       118  ...   
3              1.2              1               1       131  ...   
4              1.0              1               1       121  ...   

   triglyceride  HDL  LDL  hemoglobin  Urine protein  serum creatinine  AST  \
0           300   40   75        16.5              1               1.0   22   
1            55   57  126        16.2              1               1.1   27   
2           197   45   93        17.4              1               0.8   27   
3           203   38  102        15.9              1               1.0   20   
4            87   44   93        15.4              1               0.8   19   

   ALT  Gtp  dental caries  
0   25   27              0  
1   23   37              1  
2   31   53              0  
3   27   30              1  
4   13   17              0  

[5 rows x 23 columns]

In [3]:
#set the target
target = df["smoking"]
target.head()

0    1
1    0
2    1
3    0
4    1
Name: smoking, dtype: int64

In [6]:
#split data into training and validation
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(data.values, target.values, test_size=0.2, random_state=random_state, stratify=target)
#load the test data
#read the file
df_test = pd.read_csv("datasets/test.csv")
#get the test values
X_test = df_test.values

# Trying MLP Classifier (Neural Network)...
Results-> hidden_layer_sizes=(100, 50), max_iter=1000 Score: 0.73422

In [55]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
# Create a Multi-layer Perceptron (MLP) classifier
mlp_classifier = MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=1000, random_state=random_state)

# Fit the model on the training data
mlp_classifier.fit(X_train, y_train)

# Score it
mlp_classifier.score(X_train,y_train)


/Users/nunovivas/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:709: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


0.7290587422686886

In [54]:
#Export this prediction
# sample submission is (id,smoking)
# create an empty dataframe
df = pd.DataFrame()
pred = mlp_classifier.predict(X_test)
# add the id column
n_rows, n_cols = X_test.shape
df['id'] = df_test['id']
# add the Label column
df['smoking'] = pred
df.to_csv("submissions/submissionMLP.csv", index=False)
# show the first 5 rows
df.head()


id  smoking
0  159256        1
1  159257        0
2  159258        1
3  159259        0
4  159260        1

# Trying the forest for the first time...
Results-> Estimators: 100 Max Depth: 10 Score: 0.774
Results-> Estimators: 131 Max Depth: 10 Score: 0.767
Results-> Estimators: 091 Max Depth: 100 Score: 0.772



In [49]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

result = []
for e in range ( 1, 102, 10): # step 10
    param_grid = {
        'random_state': random_state,
        'n_estimators': e,  # The number of trees in the forest
        'max_depth': 10,      # Maximum depth of each tree
        'min_samples_split': 2,  # Minimum samples required to split an internal node
        'min_samples_leaf': 1   # Minimum samples required to be at a leaf node
        # deprecated'max_features': 'auto'   # Number of features to consider for the best split
    }
    randomForest = RandomForestClassifier(**param_grid)
    randomForest.fit(X_train, y_train)
    # predictions
    s = "N Estimators:" + str(e) + " Score: " + str(randomForest.score(X_val, y_val))
    result.append(s)
print (result)

['N Estimators:1 Score: 0.6890305161371343', 'N Estimators:11 Score: 0.7515697601406505', 'N Estimators:21 Score: 0.7631231947758382', 'N Estimators:31 Score: 0.7661685294487003', 'N Estimators:41 Score: 0.7697475825693834', 'N Estimators:51 Score: 0.7700929298003265', 'N Estimators:61 Score: 0.7705952530453347', 'N Estimators:71 Score: 0.7713173427100339', 'N Estimators:81 Score: 0.7723219892000502', 'N Estimators:91 Score: 0.7725103604169283', 'N Estimators:101 Score: 0.771662689940977']


In [43]:
# export the data
# sample submission is (id,smoking)
# create an empty dataframe
df = pd.DataFrame()
randomForest = RandomForestClassifier(n_estimators=100, random_state=random_state)
randomForest.fit(X_train, y_train)
pred = randomForest.predict(X_test)
# add the id column
n_rows, n_cols = X_test.shape
df['id'] = df_test['id']
# add the Label column
df['smoking'] = pred
df.to_csv("submissions/submissionRF.csv", index=False)
# show the first 5 rows
df.head()

id  smoking
0  159256        0
1  159257        0
2  159258        1
3  159259        0
4  159260        0

# Looper for KNN
Knn Best Score with Distance N=10 Score = 0.6684
KNN Best Score with Uniform K=15 Score = 0.6726

In [ ]:
#Looper for KNN
from sklearn.neighbors import KNeighborsClassifier
result = []
for k in range ( 1, 20):
    knn = KNeighborsClassifier(n_neighbors=k, weights='uniform')
    knn.fit(X_train, y_train)
    s = "K:" + str( k) + " Score: " + str(knn.score(X_val, y_val))
    result.append(s)
print (result)

In [26]:
#generate the submission
#read the file
df_test = pd.read_csv("datasets/test.csv")
#get the test values
X_test = df_test.values
#do predictions
knn = KNeighborsClassifier(n_neighbors=15, weights='uniform')
knn.fit(X_train, y_train)
pred = knn.predict(X_test)

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [30]:
# sample submission is (id,smoking)
# create an empty dataframe
df = pd.DataFrame()
# add the id column
n_rows, n_cols = X_test.shape
df['id'] = df_test['id']
# add the Label column
df['smoking'] = pred
df.to_csv("submissions/submissionKnn.csv", index=False)
# show the first 5 rows
df.head()

id  smoking
0  159256        0
1  159257        0
2  159258        0
3  159259        0
4  159260        1

 # Logistic Regression
 With cross validation for scoring

In [10]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

log_reg = LogisticRegression(
    random_state=random_state,
    C=100,
    solver='lbfgs',
    multi_class='auto',
    max_iter=10000)
scores = cross_val_score(estimator=log_reg, # the model
                         X=X_train, y=y_train, # X, y
                         cv=5,       #number of folds - default 5-fold cross validation (see alternatives in documentation)
                         n_jobs=-1,  # use all CPU
                         verbose=1,   # verbose level 
                         )

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   2 out of   5 | elapsed:    4.4s remaining:    6.6s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:    4.9s finished


In [12]:
scores.mean()

0.7253226478287729

In [ ]:
# Mean score : 0,72


# SVC - Support Vector Machines
## Shouldnt be executed on the Mac

In [ ]:
from sklearn.svm import SVC
#Create an SVM classifier (Support Vector Machine)
svm_classifier = SVC(kernel='linear', C=1.0, random_state=random_state,probability=True)# n_jobs not available

# Train the SVM model on the training data
svm_classifier.fit(X_train, y_train)

# Make predictions on the test data
y_pred = svm_classifier.predict_proba(X_test)

In [ ]:
# Calculate the accuracy score on the test data
accuracy = svm_classifier.score(X_test, y_train)

# Print the accuracy score
print("Accuracy Score:", accuracy)

# Decision Tree
Scores are given with an Cross Validation
Best Score = 0.69


In [11]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score,cross_val_predict
# Create a decision tree classifier
clf = DecisionTreeClassifier(criterion='gini', max_depth=None,random_state=random_state) # n_jobs not available

# Perform 5-fold cross-validation and obtain class probabilities
y_probabilities = cross_val_predict(clf, X_train, y_train, cv=5, method='predict_proba')

# Perform 5-fold cross-validation and calculate accuracy scores
scores = cross_val_score(clf, X_train, y_train, cv=5, scoring='accuracy')

# Print the accuracy scores for each fold
for i, score in enumerate(scores):
    print(f"Fold {i+1} Accuracy Score: {score:.2f}")

# Calculate and print the mean accuracy score
mean_accuracy = scores.mean()
print(f"Mean Accuracy Score: {mean_accuracy:.2f}")



Fold 1 Accuracy Score: 0.69
Fold 2 Accuracy Score: 0.69
Fold 3 Accuracy Score: 0.69
Fold 4 Accuracy Score: 0.69
Fold 5 Accuracy Score: 0.69
Mean Accuracy Score: 0.69


# K-Fold

In [13]:
#K-Fold
from sklearn.model_selection import KFold,cross_val_score
from sklearn.linear_model import LogisticRegression

# Create a logistic regression classifier with your desired hyperparameters
log_reg = LogisticRegression(max_iter=1000, C=1.0,random_state=random_state)

kfold = KFold(n_splits=3,
              shuffle=True,
              random_state=random_state)

cross_val_score(estimator=log_reg,
                X=X_train,
                y=y_train,
                cv=kfold,n_jobs=-1
                )
scores.mean()

0.6934397757013768

# Stratified K Fold

In [14]:
#Stratified K Fold
from sklearn.model_selection import StratifiedKFold
skf = StratifiedKFold()

score = cross_val_score(estimator=log_reg,
                        X=X_train,
                        y=y_train,
                        cv=skf #  int, cross-validation generator or an iterable, optional
                        )
print(score)

[0.7276402  0.71881009 0.72034065 0.7307798  0.73175039]


In [17]:
score.mean()

0.7253226478287729

# Stratified Search
With Cross Validation for scoring

In [15]:
from sklearn.model_selection import StratifiedShuffleSplit
sss = StratifiedShuffleSplit(
    n_splits=10,
    train_size=.75,
    #     test_size wil be complement of the train_size
    random_state=random_state)

score = cross_val_score(estimator=log_reg,
                        X=X_train,
                        y=y_train,
                        cv=sss)
print(score)

[0.72914508 0.71752849 0.72211234 0.72496939 0.72666478 0.72349377
 0.72270886 0.71859596 0.72352516 0.72810901]


# Start to work with ensembles with the best models so far...

MLP
Random Forest
KNN (Este retirei-o no fim pq tinha resultados bem piores)

Voting Soft parece que dá melhores resultados porque usa a média das probabilidades em vez de um "hard" vote. Alem disso tem que se usar SOFT para ter predições em formato de probabilidade

In [12]:
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier

# Create the MLP classifier
mlp_classifier = MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=1000, random_state=random_state)

# Create the Random Forest classifier
rf_classifier = RandomForestClassifier(n_estimators=91, random_state=random_state,n_jobs=-1)

# Create the KNN classifies
#knn_classifier = KNeighborsClassifier(n_neighbors=15, weights='uniform')

# Create a Voting Classifier that combines the predictions of the MLP and Random Forest classifiers
ensemble_classifier = VotingClassifier(estimators=[('mlp', mlp_classifier), ('rf', rf_classifier)], voting='soft')
# probability=True)
# Train the ensemble classifier on the training data
ensemble_classifier.fit(X_train, y_train)

# Make predictions using the ensemble classifier
y_pred = ensemble_classifier.predict_proba(X_test)

#Score
ensemble_classifier.score(X_train,y_train)

/Users/nunovivas/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:709: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


AttributeError: predict_proba is not available when voting='hard'

In [19]:
y_pred[:,1]

array([0.60100126, 0.31454335, 0.61199027, ..., 0.39004452, 0.31064334,
       0.1407514 ])

In [21]:
#Create a submission for this:
#sample submission is (id,smoking)
# create an empty dataframe
df = pd.DataFrame()
# add the id column
df['id'] = df_test['id']
# add the Label column
df['smoking'] = y_pred[:,1] # apenas a probabilidade de ser fumador
df.to_csv("submissions/submissionEssembleJustBest2_WithProba.csv", index=False) #test
# show the first 5 rows
df.head()

id   smoking
0  159256  0.601001
1  159257  0.314543
2  159258  0.611990
3  159259  0.067305
4  159260  0.576266